internal/mail/gmail/gmailauth.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information
+
+package gmail
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/smtp"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/zeebo/errs"
+)
+
+// Auth is XOAUTH2 implementation of smtp.Auth interface for gmail
+type Auth struct {
+	UserEmail string
+
+	Storage *TokenStore
+}
+
+// Start returns proto and auth credentials for first auth msg
+func (auth *Auth) Start(server *smtp.ServerInfo) (proto string, toServer []byte, err error) {
+	token, err := auth.Storage.Token()
+	if err != nil {
+		return "", nil, err
+	}
+
+	format := fmt.Sprintf("user=%s\001auth=%s %s\001\001", auth.UserEmail, token.Type, token.AccessToken)
+	return "XOAUTH2", []byte(format), nil
+}
+
+// Next sends empty response to solve SASL challenge if response code is 334
+func (auth *Auth) Next(fromServer []byte, more bool) (toServer []byte, err error) {
+	if more {
+		return make([]byte, 0), nil
+	}
+
+	return nil, nil
```

Comment:
The whole func pretty much could be `return nil, nil`, the make thing seems superfluous... Also, it looks like it should contain something more?

Reply:
that's what SASL challenge requires. If you send wrong credentials you will receive encoded error with 334 code, to solve the challenge you need to send an empty response. Making bytes array explicitly is to prevent return on nil in golang SMTP library https://github.com/golang/go/blob/e1b49ad608c0b51acf75f8248e1167acbbcb5d30/src/net/smtp/smtp.go#L233, which will allow you to send next command even thought your authentication has failed.